Sources/Foundation/Decimal.swift

```diff
@@ -1687,64 +1687,74 @@ public func NSDecimalString(_ dcm: UnsafePointer<Decimal>, _ locale: Any?) -> St
         return "0"
     }
 
+    var decimalSeparator: String!
+    if let locale = locale as? Locale {
+        decimalSeparator = locale.decimalSeparator
+    } else if let dictionary = locale as? [NSLocale.Key: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator]
+    } else if let dictionary = locale as? [String: String] {
+        decimalSeparator = dictionary[NSLocale.Key.decimalSeparator.rawValue]
+    } else {
+        decimalSeparator = "."
+    }
+    let decimalSeparatorCChar = decimalSeparator.utf8CString // should be accessed with withUnsafeBufferPointer
+    let decimalSeparatorCCharCount = decimalSeparatorCChar.withUnsafeBufferPointer { $0.count } - 1 // without null terminator
+
     let ZERO: CChar = 0x30 // ASCII '0' == 0x30
-    let DECIMALPOINT: CChar = 0x2e // ASCII '.' == 0x2e
     let MINUS: CChar = 0x2d // ASCII '-' == 0x2d
-    let bufferSize = 200 // Max value: 39 + 128 + sign + decimal point
-    var buffer = Array<CChar>(repeating: 0, count: bufferSize)
+    // FIXME: where magic constants 39 and 128 come from?
+    let bufferSize = 39 + 128 + 1 /* sign */ + decimalSeparatorCCharCount + 1 /* null terminator */
 
     var i = bufferSize - 1
-    while copy._exponent > 0 {
-        i -= 1
-        buffer[i] = ZERO
-        copy._exponent -= 1
-    }
-
-    if copy._exponent == 0 {
-        copy._exponent = 1
-    }
+    let buffer = Array<CChar>(unsafeUninitializedCapacity: bufferSize) { buf, initializedCount in
+        buf[bufferSize-1] = 0 // null terminator
+        while copy._exponent > 0 {
+            i -= 1
+            buf[i] = ZERO
+            copy._exponent -= 1
+        }
 
-    while copy._length != 0 {
-        var remainder: UInt16 = 0
         if copy._exponent == 0 {
+            copy._exponent = 1
+        }
+
+        while copy._length != 0 {
+            var remainder: UInt16 = 0
+            if copy._exponent == 0 {
+                i -= decimalSeparatorCCharCount
```

Comment:
Since the `decimalSeparator` is not always required, the initial work earlier in the function to find it could be moved into a scoped function and then only called if needed, eg could add a line above this once with:
```
let (decimalSeparatorCChar , decimalSeparatorCCharCount) = decimalSeparator()
```
Although I would probably still use the `UTF8View` as I think it might be quicker.